File: cass/apis/canvas/matching.py
"""Canvas LMS integration — roster fetching, submission retrieval, and grade push.

All raw API calls are delegated to ``CanvasClient`` in ``client.py``.
Matching logic (name normalization, student matching, slug utilities) has
been moved to ``cass.actions.matching``.
"""

from __future__ import annotations

__docformat__ = "google"

from datetime import datetime

from ...db.schema import CanvasSubmission
from .client import CanvasClient, get_token, save_token
from .schema import CanvasAssignmentResponse, CanvasStudentResponse

# Re-export for backwards compatibility (used by cli.py init)
__all__ = [
    "fetch_canvas_assignments",
    "fetch_canvas_submissions",
    "fetch_course_name",
    "fetch_students",
    "fetch_students_with_sections",
    "get_token",
    "push_grade",
    "save_token",
]


# --- API functions (delegate to CanvasClient) ---


def fetch_students(course_id: int) -> list[CanvasStudentResponse]:
    """Fetch all students enrolled in a Canvas course."""
    with CanvasClient(course_id=course_id) as c:
        return c.list_students()


def fetch_course_name(course_id: int) -> str:
    """Fetch the course name from the Canvas API."""
    with CanvasClient(course_id=course_id) as c:
        return c.get_course().name


def fetch_students_with_sections(
    course_id: int,
) -> tuple[list[CanvasStudentResponse], dict[int, str]]:
    """Fetch students and build a canvas_id -> sis_section_id mapping.

    Uses the users endpoint (with enrollments) and sections endpoint
    to resolve each student's SIS section ID.

    Returns:
        Tuple of (students, sis_section_map) where sis_section_map maps
        canvas_id to sis_section_id.
    """
    with CanvasClient(course_id=course_id) as c:
        students = c.list_students()

        # Fetch users with enrollment data to get course_section_id
        users = c.list_users(enrollment_type="student")
        user_section: dict[int, int] = {}
        for u in users:
            for e in u.enrollments:
                if e.course_section_id and e.type == "StudentEnrollment":
                    user_section[u.id] = e.course_section_id
                    break

        # Fetch sections to get sis_section_id
        sections = c.list_sections()
        section_sis: dict[int, str] = {
            s.id: s.sis_section_id for s in sections if s.sis_section_id
        }

        # Map: student canvas_id -> sis_section_id
        sis_section_map: dict[int, str] = {}
        for student in students:
            sec_id = user_section.get(student.id)
            if sec_id and sec_id in section_sis:
                sis_section_map[student.id] = section_sis[sec_id]

    return students, sis_section_map


def fetch_canvas_assignments(
    course_id: int,
) -> tuple[list[CanvasAssignmentResponse], dict[int, str]]:
    """Fetch all assignments and assignment group names from Canvas.

    Returns:
        Tuple of (assignments, group_names) where group_names maps
        assignment_group_id to group name.
    """
    with CanvasClient(course_id=course_id) as c:
        assignments = c.list_assignments()
        groups = c.list_assignment_groups()
        group_names = {g.id: g.name for g in groups}
        return assignments, group_names


def fetch_canvas_submissions(
    course_id: int,
    canvas_assignment_id: int,
    known_canvas_ids: set[int],
) -> list[CanvasSubmission]:
    """Fetch submissions for a Canvas assignment, filtered to known students.

    Args:
        course_id: Canvas course ID.
        canvas_assignment_id: Canvas assignment ID.
        known_canvas_ids: Set of canvas_ids from the students table.

    Returns:
        CanvasSubmission domain objects for known students.
    """
    with CanvasClient(course_id=course_id) as c:
        raw = c.list_submissions(canvas_assignment_id)

    submissions: list[CanvasSubmission] = []
    for r in raw:
        if r.user_id not in known_canvas_ids:
            continue

        submitted_at = None
        if r.submitted_at:
            try:
                submitted_at = datetime.fromisoformat(r.submitted_at)
            except ValueError:
                pass

        submissions.append(
            CanvasSubmission(
                canvas_user_id=r.user_id,
                canvas_assignment_id=canvas_assignment_id,
                submitted=r.workflow_state in ("submitted", "graded", "pending_review"),
                submitted_at=submitted_at,
                late=r.late,
                lateness_seconds=int(r.seconds_late),
                score=r.score,
                workflow_state=r.workflow_state,
            )
        )

    return submissions


def push_grade(
    course_id: int, assignment_id: int, student_canvas_id: int, grade: str
) -> bool:
    """Push a single grade to Canvas. Returns True on success, False on failure."""
    with CanvasClient(course_id=course_id) as c:
        return c.push_grade(assignment_id, student_canvas_id, grade)

File: src/apis/canvas/matching.ts
/**
 * Canvas roster and submission fetching — transforms API responses to DB rows.
 */
import type { KyInstance } from "ky";
import type { NewCanvasAssignment, NewCanvasSubmission, NewStudent } from "../../db/schema.ts";
import { getPaginated } from "./client.ts";
import {
	CanvasAssignmentGroup,
	CanvasAssignmentResponse,
	CanvasSection,
	CanvasStudentResponse,
	CanvasSubmissionResponse,
	CanvasUser,
} from "./schema.ts";

const SUBMITTED_STATES = new Set(["submitted", "graded", "pending_review"]);

/** Fetch students with section info. Returns [students, sisSectionMap]. */
export async function fetchStudentsWithSections(
	client: KyInstance,
	courseId: number,
): Promise<[NewStudent[], Map<number, string>]> {
	const [rawStudents, rawUsers, rawSections] = await Promise.all([
		getPaginated<unknown>(client, `courses/${courseId}/users`, {
			enrollment_type: "student",
		}).then((items) => items.map((i) => CanvasStudentResponse.parse(i))),

		getPaginated<unknown>(client, `courses/${courseId}/users`, {
			enrollment_type: "student",
			include: "enrollments",
		}).then((items) => items.map((i) => CanvasUser.parse(i))),

		getPaginated<unknown>(client, `courses/${courseId}/sections`).then((items) =>
			items.map((i) => CanvasSection.parse(i)),
		),
	]);

	// Build section map: section_id -> sis_section_id
	const sectionMap = new Map<number, string>();
	for (const s of rawSections) {
		if (s.sis_section_id) sectionMap.set(s.id, s.sis_section_id);
	}

	// Build user -> section_id from enrollments
	const userSectionMap = new Map<number, number>();
	for (const u of rawUsers) {
		const studentEnrollment = u.enrollments.find((e) => e.type === "StudentEnrollment");
		if (studentEnrollment) {
			userSectionMap.set(u.id, studentEnrollment.course_section_id);
		}
	}

	// Build sis_section_id map for each canvas_id
	const sisSectionMap = new Map<number, string>();
	for (const [userId, sectionId] of userSectionMap) {
		const sis = sectionMap.get(sectionId);
		if (sis) sisSectionMap.set(userId, sis);
	}

	const students: NewStudent[] = rawStudents.map((s) => ({
		canvas_id: s.id,
		name: s.name,
		sortable_name: s.sortable_name,
		email: s.email ?? "",
		login_id: s.login_id ?? "",
		sis_user_id: s.sis_user_id ?? "",
		sis_section_id: sisSectionMap.get(s.id) ?? "",
		github_username: null,
	}));

	return [students, sisSectionMap];
}

/** Fetch canvas assignments with group names. Returns [assignments, groupNames]. */
export async function fetchCanvasAssignments(
	client: KyInstance,
	courseId: number,
): Promise<[NewCanvasAssignment[], Map<number, string>]> {
	const [rawAssignments, rawGroups] = await Promise.all([
		getPaginated<unknown>(client, `courses/${courseId}/assignments`).then((items) =>
			items.map((i) => CanvasAssignmentResponse.parse(i)),
		),
		getPaginated<unknown>(client, `courses/${courseId}/assignment_groups`).then((items) =>
			items.map((i) => CanvasAssignmentGroup.parse(i)),
		),
	]);

	const groupNames = new Map<number, string>();
	for (const g of rawGroups) {
		groupNames.set(g.id, g.name);
	}

	const assignments: NewCanvasAssignment[] = rawAssignments.map((a) => ({
		canvas_id: a.id,
		name: a.name,
		points_possible: a.points_possible ?? 0,
		due_at: a.due_at,
		published: a.published ? 1 : 0,
		assignment_group: a.assignment_group_id ? (groupNames.get(a.assignment_group_id) ?? "") : "",
		post_manually: a.post_manually ? 1 : 0,
	}));

	return [assignments, groupNames];
}

/** Fetch submissions for a single assignment. */
export async function fetchCanvasSubmissions(
	client: KyInstance,
	courseId: number,
	canvasAssignmentId: number,
	knownCanvasIds: Set<number>,
): Promise<NewCanvasSubmission[]> {
	const rawSubs = await getPaginated<unknown>(
		client,
		`courses/${courseId}/assignments/${canvasAssignmentId}/submissions`,
	).then((items) => items.map((i) => CanvasSubmissionResponse.parse(i)));

	const now = Date.now() / 1000;
	return rawSubs
		.filter((s) => knownCanvasIds.has(s.user_id))
		.map((s) => ({
			canvas_user_id: s.user_id,
			canvas_assignment_id: canvasAssignmentId,
			submitted: SUBMITTED_STATES.has(s.workflow_state) ? 1 : 0,
			submitted_at: s.submitted_at,
			late: s.late ? 1 : 0,
			lateness_seconds: Math.floor(s.seconds_late),
			score: s.score,
			workflow_state: s.workflow_state,
			fetched_at: now,
			posted_grade: s.grade ?? "",
			grade_updated_at: now,
		}));
}
